contracts/InterestRateModels/CheckpointRateModel.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.25;
+
+import { ensureNonzeroAddress } from "@venusprotocol/solidity-utilities/contracts/validators.sol";
+import { InterestRateModel } from "./InterestRateModel.sol";
+
+/**
+ * @title Venus CheckpointRateModel Contract
+ * @notice A contract that applies a rate model based on whether a certain checkpoint
+ *   in time has passed. Useful to adjust interest rates for block time changes in
+ *   the underlying networks, if the block time update is scheduled in advance.
+ * @author Venus
+ */
+contract CheckpointRateModel is InterestRateModel {
+    InterestRateModel public immutable OLD_RATE_MODEL;
+    InterestRateModel public immutable NEW_RATE_MODEL;
+    uint256 public immutable CHECKPOINT_TIMESTAMP;
+
+    /**
+     * @notice Constructor
+     * @param oldRateModel Rate model to use before the checkpoint
+     * @param newRateModel Rate model to use after the checkpoint
+     * @param checkpointTimestamp Checkpoint timestamp
+     */
+    constructor(InterestRateModel oldRateModel, InterestRateModel newRateModel, uint256 checkpointTimestamp) {
+        ensureNonzeroAddress(address(oldRateModel));
+        ensureNonzeroAddress(address(newRateModel));
+        OLD_RATE_MODEL = oldRateModel;
+        NEW_RATE_MODEL = newRateModel;
+        CHECKPOINT_TIMESTAMP = checkpointTimestamp;
+    }
+
+    /**
+     * @notice Returns the current rate model contract (either the old one or the new one)
+     * @return Rate model in use
+     */
+    function currentRateModel() external view returns (InterestRateModel) {
+        return _getCurrentRateModel();
+    }
+
+    /**
+     * @inheritdoc InterestRateModel
+     */
+    function getBorrowRate(
+        uint256 cash,
+        uint256 borrows,
+        uint256 reserves,
+        uint256 badDebt
+    ) external view override returns (uint256) {
+        return _getCurrentRateModel().getBorrowRate(cash, borrows, reserves, badDebt);
+    }
+
+    /**
+     * @inheritdoc InterestRateModel
+     */
+    function getSupplyRate(
+        uint256 cash,
+        uint256 borrows,
+        uint256 reserves,
+        uint256 reserveFactorMantissa,
+        uint256 badDebt
+    ) external view override returns (uint256) {
+        return _getCurrentRateModel().getSupplyRate(cash, borrows, reserves, reserveFactorMantissa, badDebt);
+    }
+
```

Comment:
I would also implement `utilizationRate`, used by the UI, AFAIK.

Reply:
Should we put `utilizationRate` to `InterestRateModel` interface in this case? So that all rate models implement it

Reply:
Theoretically, there could be an interest rate model not based on the utilization rate. But if that would simplify the integration with third parties (i.e. the dapp), it lgtm

Reply:
I added a separate interface (0e67188ec7968ea854159997e35e41db112b981a) and the support for utilizatoin rate (9b9a5d814b5631b4e2ccc82a00b75565638b0bfb). Let me know if you think this is the right direction. I'll also raise an alternative PR based on fallback – it seems a bit more universal and it is less likely to break any clients.